python/lsst/rubintv/production/rubinTv.py

```diff
@@ -554,3 +631,161 @@ def run(self):
+
+        currTime = int(time.time())
+        nextExpected = currTime + flatlinePeriod
+
+        heartbeatJsonDict = {
+            "channel": channel,
+            "currTime": currTime,
+            "nextExpected": nextExpected,
+            "errors": {}
+        }
+        heartbeatJson = json.dumps(heartbeatJsonDict)
+
+        blob = self.bucket.blob("/".join([filename]))
+        blob.cache_control = 'no-store'  # must set before upload
+
+        # heartbeat retry strategy
+        modified_retry = DEFAULT_RETRY.with_deadline(0.6)  # single retry here
+        modified_retry = modified_retry.with_delay(initial=0.5, multiplier=1.2, maximum=2)
+
+        try:
+            blob.upload_from_string(heartbeatJson, retry=modified_retry)
+            self.log.debug(f'Uploaded heartbeat to channel {channel} with datetime {currTime}')
+            return True
+        except Exception:
+            return False
+
+    def googleUpload(self, channel, sourceFilename, uploadAsFilename=None, isLiveFile=False):
+        """Upload a file to the RubinTV Google cloud storage bucket.
+
+        Parameters
+        ----------
+        channel : `str`
+            The RubinTV channel to upload to.
+        sourceFilename : `str`
+            The full path and filename of the file to upload.
+        uploadAsFilename : `str`, optional
+            Optionally rename the file to this upon upload.
+        isLiveFile : `bool`, optional
+            The file is being updated constantly, and so caching should be
+        disabled.
+
+        Raises
+        ------
+        ValueError
+            Raised if the specified channel is not in the list of existing
+            channels as specified in CHANNELS
+        RuntimeError
+            Raised if the Google cloud storage is not installed/importable.
+        """
+        if channel not in CHANNELS:
+            raise ValueError(f"Error: {channel} not in {CHANNELS}")
+
+        if uploadAsFilename and (os.path.basename(sourceFilename) != uploadAsFilename):
+            tempdir = tempfile.mkdtemp()
+            finalName = os.path.join(tempdir, uploadAsFilename)
+            shutil.copy(sourceFilename, finalName)
+            assert os.path.exists(finalName)
+        else:
+            finalName = sourceFilename
+
+        path = Path(finalName)
+        blob = self.bucket.blob("/".join([channel, path.name]))
+        if isLiveFile:
+            blob.cache_control = 'no-store'
+        self.log.info(f'Uploaded {sourceFilename} to {finalName}')
+
+        # general retry strategy
+        # still quite gentle as the catchup service will fill in gaps
+        # and we don't want to hold up new images landing
+        modified_retry = DEFAULT_RETRY.with_deadline(2.0)  # in seconds
+        modified_retry = modified_retry.with_delay(initial=.5, multiplier=1.2, maximum=2)
+        try:
+            blob.upload_from_filename(finalName, retry=modified_retry)
+        except Exception:
+            return None
```

Comment:
I feel like I may have asked this previously on another ticket: do you not care about the exception because this will try again later?  Just wondering if a log message for either case is helpful.

Reply:
Correct, the catchup service will try again later, but yes, I agree, no harm in logging a warning, I'll add one now.